src/features/feature_selector.cpp

```diff
@@ -0,0 +1,241 @@
+/**
+ * @file feature_selector.cpp
+ * @author Sean Massung
+ */
+
+#include <iostream>
+#include <fstream>
+#include <unordered_set>
+#include "util/filesystem.h"
+#include "util/progress.h"
+#include "parallel/parallel_for.h"
+#include "features/feature_selector.h"
+#include "index/postings_data.h"
+#include "io/binary.h"
+
+namespace meta
+{
+namespace features
+{
+feature_selector::feature_selector(const std::string& prefix,
+                                   std::shared_ptr<index::forward_index> idx)
+    : prefix_{prefix},
+      idx_{std::move(idx)},
+      selected_{prefix_ + ".selected", idx_->unique_terms()}
+{ /* nothing */
+}
+
+void feature_selector::init(uint64_t features_per_class)
+{
+    // if the first class distribution doesn't exist, we haven't created the
+    // data for this feature_selector yet
+    if (!filesystem::file_exists(prefix_ + ".1"))
+    {
+        // initially set all probabilities to zero; this allows fast random
+        // access to the probabilities
+        term_prob_.assign(idx_->unique_terms(), 0.0);
+        class_prob_.assign(idx_->num_labels(), 0.0);
+        co_occur_.assign(idx_->num_labels(),
+                         std::vector<double>(idx_->unique_terms(), 0.0));
+        calc_probs();
+        score_all();
+        select(features_per_class);
+    }
+}
+
+void feature_selector::score_all()
+{
+    using pair_t = std::pair<term_id, double>;
+    std::vector<std::vector<pair_t>> scores(
+        class_prob_.size(), std::vector<pair_t>(term_prob_.size()));
+
+    printing::progress prog{" > Selecting features: ", term_prob_.size()};
+    for (uint64_t tid = 0; tid < idx_->unique_terms(); ++tid)
+    {
+        prog(tid);
+        for (uint64_t lbl = 0; lbl < idx_->num_labels(); ++lbl)
+            scores[lbl][tid] = std::make_pair(
+                tid, score(static_cast<label_id>(lbl + 1), term_id{tid}));
+    }
+    prog.end();
+
+    parallel::parallel_for(
+        scores.begin(), scores.end(), [&](std::vector<pair_t>& v)
+        {
+            std::sort(v.begin(), v.end(), [&](const pair_t& a, const pair_t& b)
```

Comment:
The capture group can be empty here since it doesn't refer to anything.

I don't think it matters, but it could?